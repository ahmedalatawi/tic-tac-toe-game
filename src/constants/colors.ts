export const Colors = {
  gray: "#a9a9a9",
  grayLight: "#DDDDDD",
  white: "#FFFFFF",
  black: "#000000",
};
