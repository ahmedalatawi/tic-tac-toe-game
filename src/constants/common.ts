export const BOARD_SIZES = {
  _3x3: 3,
  _4x4: 4,
  _5x5: 5,
};

export const PLAYER_X = 1;
export const PLAYER_O = 2;
export const TIE = 0;

export const PLAYER_TYPES = {
  user: "user",
  computer: "computer",
};

export const GAME_STATES = {
  notStarted: "not_started",
  inProgress: "in_progress",
  ended: "ended",
};
