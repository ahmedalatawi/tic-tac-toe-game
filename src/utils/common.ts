import {
  TIE,
  PLAYER_O,
  PLAYER_X,
  PLAYER_TYPES,
  BOARD_SIZES,
} from "../constants/common";
import { GameBoard, GameSettings } from "../types/types";

export const createBoard = (size: number) =>
  new Array(size).fill(Array(size).fill(null));

// Get a random number between min and max
export const getRandomNumber = (min: number, max: number) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1)) + min;
};

export const switchPlayer = (player: number) => {
  return player === PLAYER_X ? PLAYER_O : PLAYER_X;
};

/**
 * This function is used to return a random move for the computer
 * @param board
 * @param boardSize
 * @returns
 */
export const getRandomMove = (board: GameBoard, boardSize?: number | null) => {
  const max = (boardSize ?? BOARD_SIZES._3x3) - 1;
  let row = getRandomNumber(0, max);
  let col = getRandomNumber(0, max);

  while (board[row][col]) {
    row = getRandomNumber(0, max);
    col = getRandomNumber(0, max);
  }

  return [row, col];
};

/**
 * This function returns the game's final status based on player X and O, or tie
 * @param winner
 * @returns
 */
export const getWinnerText = (winner: number) => {
  let winnerText = "";
  switch (winner) {
    case PLAYER_X:
      winnerText = "The winner is player X";
      break;
    case PLAYER_O:
      winnerText = "The winner is player O";
      break;
    case TIE:
    default:
      winnerText = "It's a tie";
  }

  return winnerText;
};

/**
 * This function calculates all winning possibilities, e.g. column, row, or diagonal in order to determine the winner
 *
 * TODO: This function can def be improved by calculating all possibilities dynamically - no hard coding
 * @param board
 * @param boardSize
 * @returns
 */
export const getWinner = (board: GameBoard, boardSize?: number | null) => {
  const dimensions = boardSize ?? BOARD_SIZES._3x3;

  const rowMarked = (row: number, board: GameBoard) => {
    if (dimensions === BOARD_SIZES._3x3) {
      return (
        board[row][0] !== null &&
        board[row][0] === board[row][1] &&
        board[row][1] === board[row][2]
      );
    }
    if (dimensions === BOARD_SIZES._4x4) {
      return (
        board[row][0] !== null &&
        board[row][0] === board[row][1] &&
        board[row][1] === board[row][2] &&
        board[row][2] === board[row][3]
      );
    }
    if (dimensions === BOARD_SIZES._5x5) {
      return (
        board[row][0] !== null &&
        board[row][0] === board[row][1] &&
        board[row][1] === board[row][2] &&
        board[row][2] === board[row][3] &&
        board[row][3] === board[row][4]
      );
    }
  };

  const colMarked = (column: number, board: GameBoard) => {
    if (dimensions === BOARD_SIZES._3x3) {
      return (
        board[0][column] !== null &&
        board[0][column] === board[1][column] &&
        board[1][column] === board[2][column]
      );
    }
    if (dimensions === BOARD_SIZES._4x4) {
      return (
        board[0][column] !== null &&
        board[0][column] === board[1][column] &&
        board[1][column] === board[2][column] &&
        board[2][column] === board[3][column]
      );
    }
    if (dimensions === BOARD_SIZES._5x5) {
      return (
        board[0][column] !== null &&
        board[0][column] === board[1][column] &&
        board[1][column] === board[2][column] &&
        board[2][column] === board[3][column] &&
        board[3][column] === board[4][column]
      );
    }
  };

  const diagonalMarked = (diagonal: number, board: GameBoard) => {
    if (dimensions === BOARD_SIZES._3x3) {
      if (diagonal === 1) {
        return (
          board[0][0] !== null &&
          board[0][0] === board[1][1] &&
          board[1][1] === board[2][2]
        );
      } else {
        return (
          board[2][0] !== null &&
          board[2][0] === board[1][1] &&
          board[1][1] === board[0][2]
        );
      }
    }
    if (dimensions === BOARD_SIZES._4x4) {
      if (diagonal === 1) {
        return (
          board[0][0] !== null &&
          board[0][0] === board[1][1] &&
          board[1][1] === board[2][2] &&
          board[2][2] === board[3][3]
        );
      } else {
        return (
          board[2][0] !== null &&
          board[3][0] === board[2][1] &&
          board[2][1] === board[1][2] &&
          board[1][2] === board[0][3]
        );
      }
    }
    if (dimensions === BOARD_SIZES._5x5) {
      if (diagonal === 1) {
        return (
          board[0][0] !== null &&
          board[0][0] === board[1][1] &&
          board[1][1] === board[2][2] &&
          board[2][2] === board[3][3] &&
          board[3][3] === board[4][4]
        );
      } else {
        return (
          board[2][0] !== null &&
          board[4][0] === board[3][1] &&
          board[3][1] === board[2][2] &&
          board[2][2] === board[1][3] &&
          board[1][3] === board[0][4]
        );
      }
    }
  };

  if (dimensions === BOARD_SIZES._5x5 && rowMarked(4, board)) {
    return board[4][0];
  }
  if (dimensions === BOARD_SIZES._4x4 && rowMarked(3, board)) {
    return board[3][0];
  }

  if (dimensions === BOARD_SIZES._5x5 && colMarked(4, board)) {
    return board[0][4];
  }
  if (dimensions === BOARD_SIZES._4x4 && colMarked(3, board)) {
    return board[0][3];
  }

  if (dimensions === BOARD_SIZES._5x5 && diagonalMarked(2, board)) {
    return board[4][0];
  }
  if (dimensions === BOARD_SIZES._4x4 && diagonalMarked(2, board)) {
    return board[3][0];
  }

  if (rowMarked(0, board)) {
    return board[0][0];
  }
  if (rowMarked(1, board)) {
    return board[1][0];
  }
  if (rowMarked(2, board)) {
    return board[2][0];
  }

  if (colMarked(0, board)) {
    return board[0][0];
  }
  if (colMarked(1, board)) {
    return board[0][1];
  }
  if (colMarked(2, board)) {
    return board[0][2];
  }

  if (diagonalMarked(1, board)) {
    return board[0][0];
  }
  if (diagonalMarked(2, board)) {
    return board[2][0];
  }

  // if we get here, we need to check for any empty spot before we determine if game is a tie
  for (let row = 0; row < board.length; row++) {
    for (let col = 0; col < board[row].length; col++) {
      if (board[row][col] === null) {
        return null;
      }
    }
  }

  return TIE;
};

export const isUserVsUser = (settings: GameSettings) =>
  settings.playerX === PLAYER_TYPES.user &&
  settings.playerO === PLAYER_TYPES.user;

export const isComputerVsComputer = (settings: GameSettings) =>
  settings.playerX === PLAYER_TYPES.computer &&
  settings.playerO === PLAYER_TYPES.computer;

export const getUserValue = (settings: GameSettings) =>
  isUserVsUser(settings)
    ? PLAYER_X
    : settings.playerX === PLAYER_TYPES.user
    ? PLAYER_X
    : PLAYER_O;

export const getComputerValue = (settings: GameSettings) =>
  isComputerVsComputer(settings)
    ? PLAYER_X
    : settings.playerO === PLAYER_TYPES.computer
    ? PLAYER_O
    : PLAYER_X;
