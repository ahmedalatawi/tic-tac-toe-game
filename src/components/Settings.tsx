import {
  View,
  StyleSheet,
  TouchableOpacity,
  Text,
  Platform,
} from "react-native";
import { BOARD_SIZES, PLAYER_TYPES } from "../constants/common";
import React, { useEffect, useState } from "react";
import { Picker } from "@react-native-picker/picker";
import { GameSettings, PlayerTypes } from "../types/types";
import RadioButton from "./ui/RadioButton";
import { Colors } from "../constants/colors";
import Checkbox from "expo-checkbox";

interface Props {
  onStartGame: (settings: GameSettings) => void;
}

type GameSettingsKeys = keyof GameSettings;

function Settings({ onStartGame }: Props) {
  const [settings, setSettings] = useState<GameSettings>({
    playerX: "user",
    playerO: "computer",
    useAi: false,
    boardSize: BOARD_SIZES._3x3,
  });
  const [isUseAiChecked, setIsUseAiChecked] = useState(false);

  const handleBoardSizePress = (value: number) => {
    setSettings((currentSettings) => ({
      ...currentSettings,
      boardSize: value,
    }));
  };

  const handleAiCheckboxPress = (value: boolean) => {
    setIsUseAiChecked(value);
    setSettings((currentSettings) => ({
      ...currentSettings,
      useAi: value,
    }));
  };

  const handlePlayerDropdownPress = (
    value: PlayerTypes | null,
    key: GameSettingsKeys
  ) => {
    setSettings((currentSettings) => ({
      ...currentSettings,
      [key]: value,
    }));
  };

  const disableStartGameBtn = !settings.playerX || !settings.playerO;

  const disableAiCheckbox =
    (settings.playerX !== PLAYER_TYPES.computer &&
      settings.playerO !== PLAYER_TYPES.computer) ||
    //TODO: Currently AI is only available for 3x3 board.
    //There seems to be a performance issue that might cause your device (or virtual device)
    //to freeze when using the AI Algorithm for other board sizes.
    settings.boardSize !== BOARD_SIZES._3x3;

  useEffect(() => {
    disableAiCheckbox && setIsUseAiChecked(false);
  }, [disableAiCheckbox]);

  return (
    <View style={styles.container}>
      <View style={styles.settingsContainer}>
        <View style={styles.boardSizesContainer}>
          <View style={styles.radioBtnContainer}>
            <RadioButton
              testID="3x3TestId"
              selected={settings.boardSize === BOARD_SIZES._3x3}
              onSelect={() => handleBoardSizePress(BOARD_SIZES._3x3)}
            />
            <Text style={styles.radioBtnText}>3x3</Text>
          </View>
          <View style={styles.radioBtnContainer}>
            <RadioButton
              testID="4x4TestId"
              selected={settings.boardSize === BOARD_SIZES._4x4}
              onSelect={() => handleBoardSizePress(BOARD_SIZES._4x4)}
            />
            <Text style={styles.radioBtnText}>4x4</Text>
          </View>
          <View style={styles.radioBtnContainer}>
            <RadioButton
              testID="5x5TestId"
              selected={settings.boardSize === BOARD_SIZES._5x5}
              onSelect={() => handleBoardSizePress(BOARD_SIZES._5x5)}
            />
            <Text style={styles.radioBtnText}>5x5</Text>
          </View>
        </View>
        <View style={styles.checkboxContainer}>
          <Checkbox
            testID="aiCheckboxTestId"
            style={styles.checkbox}
            value={isUseAiChecked}
            disabled={disableAiCheckbox}
            onValueChange={handleAiCheckboxPress}
          />
          <Text style={styles.checkboxText}>Incorporate AI</Text>
        </View>
        <Text style={styles.selectPlayersText}>
          Select player "X" and "O" to start the game
        </Text>
        <View style={styles.pickersContainer}>
          <View style={[styles.pickerContainer, getAndroidPickerStyles()]}>
            <Picker<PlayerTypes | null>
              style={styles.picker}
              mode="dropdown"
              selectedValue={settings.playerX}
              onValueChange={(itemValue) =>
                handlePlayerDropdownPress(itemValue, "playerX")
              }
            >
              {Platform.OS === "android" && (
                <Picker.Item
                  value={null}
                  label="-- Player X --"
                  style={styles.pickerItem}
                />
              )}
              <Picker.Item label="User" value={PLAYER_TYPES.user} />
              <Picker.Item label="Computer" value={PLAYER_TYPES.computer} />
            </Picker>
          </View>
          <View style={[styles.pickerContainer, getAndroidPickerStyles()]}>
            <Picker<PlayerTypes | null>
              style={styles.picker}
              mode="dropdown"
              selectedValue={settings.playerO}
              onValueChange={(itemValue) =>
                handlePlayerDropdownPress(itemValue, "playerO")
              }
            >
              {Platform.OS === "android" && (
                <Picker.Item
                  value={null}
                  label="-- Player O --"
                  style={styles.pickerItem}
                />
              )}
              <Picker.Item label="User" value={PLAYER_TYPES.user} />
              <Picker.Item label="Computer" value={PLAYER_TYPES.computer} />
            </Picker>
          </View>
        </View>
      </View>
      <View style={styles.startGameBtnContainer}>
        <TouchableOpacity
          testID="startGameBtnTestId"
          style={styles.selectorBtn}
          onPress={() => onStartGame(settings)}
          disabled={disableStartGameBtn}
        >
          <Text
            style={[
              styles.startGameText,
              { color: disableStartGameBtn ? Colors.gray : Colors.black },
            ]}
          >
            Start game
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  );
}

const getAndroidPickerStyles = () =>
  Platform.OS === "android" && {
    borderRadius: 10,
    borderWidth: 1,
    borderColor: Colors.gray,
  };

const styles = StyleSheet.create({
  container: {
    rowGap: 32,
  },
  boardSizesContainer: {
    flexDirection: "row",
    gap: 20,
    alignItems: "center",
  },
  settingsContainer: {
    alignItems: "center",
    paddingHorizontal: 15,
  },
  selectPlayersText: {
    fontSize: 16,
    paddingBottom: Platform.OS === "ios" ? 70 : 10,
    paddingTop: 20,
  },
  selectorBtn: {
    alignItems: "center",
    backgroundColor: Colors.grayLight,
    padding: 10,
    margin: 10,
    width: "100%",
    borderRadius: 5,
  },
  startGameText: {
    fontSize: 18,
  },
  pickersContainer: {
    flexDirection: "row",
    gap: 10,
  },
  pickerContainer: {
    height: 40,
    flexDirection: "row",
    alignItems: "center",
    width: "50%",
  },
  picker: {
    width: "100%",
  },
  pickerItem: {
    color: "gray",
  },
  radioBtnContainer: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    paddingVertical: 20,
    padding: 20,
    gap: 5,
  },
  radioBtnText: {
    fontSize: 18,
  },
  startGameBtnContainer: {
    justifyContent: "center",
    alignItems: "center",
    paddingHorizontal: 15,
    paddingTop: 8,
  },
  checkboxContainer: {
    marginHorizontal: 16,
    marginVertical: 20,
    flexDirection: "row",
    alignItems: "center",
  },
  section: {
    flexDirection: "row",
    alignItems: "center",
  },
  paragraph: {
    fontSize: 15,
  },
  checkbox: {
    margin: 8,
  },
  checkboxText: {
    fontSize: 15,
  },
});

export default Settings;
