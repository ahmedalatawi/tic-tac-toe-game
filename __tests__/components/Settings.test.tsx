import React from "react";
import renderer from "react-test-renderer";
import { render, screen, userEvent } from "@testing-library/react-native";
import { BOARD_SIZES } from "../../src/constants/common";
import Settings from "../../src/components/Settings";

const settingsMock = {
  playerX: "user",
  playerO: "computer",
  useAi: false,
  boardSize: BOARD_SIZES._3x3,
};

describe("<Settings />", () => {
  test("renders correctly", () => {
    const onStartGameMock = jest.fn();
    const tree = renderer
      .create(<Settings onStartGame={onStartGameMock} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("should emit the correct settings", async () => {
    const onStartGameMock = jest.fn();
    render(<Settings onStartGame={onStartGameMock} />);

    const board3x3 = screen.getByTestId("3x3TestId");
    const board4x4 = screen.getByTestId("4x4TestId");
    const board5x5 = screen.getByTestId("5x5TestId");

    expect(board3x3).toBeDefined();

    const startGameBtn = screen.getByTestId("startGameBtnTestId");

    await userEvent.press(startGameBtn);

    //default settings
    expect(onStartGameMock).toHaveBeenCalledWith(settingsMock);

    //select board 4x4
    await userEvent.press(board4x4);
    await userEvent.press(startGameBtn);

    expect(onStartGameMock).toHaveBeenCalledWith({
      ...settingsMock,
      boardSize: BOARD_SIZES._4x4,
    });

    //select board 5x5
    await userEvent.press(board5x5);
    await userEvent.press(startGameBtn);

    expect(onStartGameMock).toHaveBeenCalledWith({
      ...settingsMock,
      boardSize: BOARD_SIZES._5x5,
    });
  });

  test("should check the AI checkbox", async () => {
    const onStartGameMock = jest.fn();
    render(<Settings onStartGame={onStartGameMock} />);

    const aiCheckbox = screen.getByTestId("aiCheckboxTestId");

    await userEvent.press(aiCheckbox);

    const startGameBtn = screen.getByTestId("startGameBtnTestId");

    await userEvent.press(startGameBtn);

    expect(onStartGameMock).toHaveBeenCalledWith({
      ...settingsMock,
      useAi: true,
    });
  });

  //FIXME: needs fixed!
  xtest("should select the right player from dropdown", async () => {
    const onStartGameMock = jest.fn();
    render(<Settings onStartGame={onStartGameMock} />);

    const computerPlayer = screen.getByText("User");

    await userEvent.press(computerPlayer);

    const startGameBtn = screen.getByTestId("startGameBtnTestId");

    await userEvent.press(startGameBtn);

    expect(onStartGameMock).toHaveBeenCalledWith({
      ...settingsMock,
      playerX: "computer",
    });
  });
});
