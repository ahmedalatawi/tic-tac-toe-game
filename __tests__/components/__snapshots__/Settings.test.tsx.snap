// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> renders correctly 1`] = `
<View
  style={
    {
      "rowGap": 32,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "paddingHorizontal": 15,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 20,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 5,
            "justifyContent": "center",
            "padding": 10,
            "paddingVertical": 20,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "black",
              "borderRadius": 14,
              "borderWidth": 2,
              "height": 18,
              "justifyContent": "center",
              "opacity": 1,
              "width": 18,
            }
          }
          testID="3x3TestId"
        >
          <View
            style={
              {
                "backgroundColor": "black",
                "borderRadius": 7,
                "height": 10,
                "width": 10,
              }
            }
          />
        </View>
        <Text
          style={
            {
              "fontSize": 18,
            }
          }
        >
          3x3
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 5,
            "justifyContent": "center",
            "padding": 10,
            "paddingVertical": 20,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "black",
              "borderRadius": 14,
              "borderWidth": 2,
              "height": 18,
              "justifyContent": "center",
              "opacity": 1,
              "width": 18,
            }
          }
          testID="4x4TestId"
        />
        <Text
          style={
            {
              "fontSize": 18,
            }
          }
        >
          4x4
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 5,
            "justifyContent": "center",
            "padding": 10,
            "paddingVertical": 20,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "borderColor": "black",
              "borderRadius": 14,
              "borderWidth": 2,
              "height": 18,
              "justifyContent": "center",
              "opacity": 1,
              "width": 18,
            }
          }
          testID="5x5TestId"
        />
        <Text
          style={
            {
              "fontSize": 18,
            }
          }
        >
          5x5
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginHorizontal": 16,
          "marginVertical": 10,
        }
      }
    >
      <View
        accessibilityRole="checkbox"
        accessibilityState={
          {
            "busy": undefined,
            "checked": false,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "borderColor": "#657786",
              "borderRadius": 2,
              "borderWidth": 2,
              "height": 20,
              "width": 20,
            },
            {
              "margin": 8,
            },
            false,
            false,
            false,
            false,
          ]
        }
        testID="aiCheckboxTestId"
      />
      <Text
        style={
          {
            "fontSize": 15,
          }
        }
      >
        Incorporate AI
      </Text>
    </View>
    <Text
      style={
        {
          "fontSize": 16,
          "paddingBottom": 10,
          "paddingTop": 20,
        }
      }
    >
      Select player "X" and "O" to start the game
    </Text>
    <View
      style={
        {
          "flexDirection": "row",
          "gap": 10,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderColor": "#a9a9a9",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 40,
            "width": "50%",
          }
        }
      >
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <RNCPicker
            items={
              [
                {
                  "label": "-- Player X --",
                  "testID": undefined,
                  "textColor": undefined,
                  "value": "null",
                },
                {
                  "label": "User",
                  "testID": undefined,
                  "textColor": undefined,
                  "value": "user",
                },
                {
                  "label": "Computer",
                  "testID": undefined,
                  "textColor": undefined,
                  "value": "computer",
                },
              ]
            }
            numberOfLines={1}
            onChange={[Function]}
            selectedIndex={1}
            style={
              [
                {
                  "height": 216,
                },
                undefined,
              ]
            }
          />
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "borderColor": "#a9a9a9",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 40,
            "width": "50%",
          }
        }
      >
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <RNCPicker
            items={
              [
                {
                  "label": "-- Player O --",
                  "testID": undefined,
                  "textColor": undefined,
                  "value": "null",
                },
                {
                  "label": "User",
                  "testID": undefined,
                  "textColor": undefined,
                  "value": "user",
                },
                {
                  "label": "Computer",
                  "testID": undefined,
                  "textColor": undefined,
                  "value": "computer",
                },
              ]
            }
            numberOfLines={1}
            onChange={[Function]}
            selectedIndex={2}
            style={
              [
                {
                  "height": 216,
                },
                undefined,
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "justifyContent": "center",
        "paddingHorizontal": 15,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#DDDDDD",
          "borderRadius": 5,
          "margin": 10,
          "opacity": 1,
          "padding": 10,
          "width": "100%",
        }
      }
      testID="startGameBtnTestId"
    >
      <Text
        style={
          [
            {
              "fontSize": 18,
            },
            {
              "color": "#000000",
            },
          ]
        }
      >
        Start game
      </Text>
    </View>
  </View>
</View>
`;
